Tools overview

TGWDB opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [V ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

martin-cs commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,228 @@
+- [others (converter, driver, file_converter, java-converter)](#others)
+
+The rest of this document provides a section on each of these tools in alphabetical order.
+Most links in this document are to the [CProver online documentation](http://cprover.diffblue.com/index.html).
```
I'm not quite sure how linking is done in the rest of the documentation but if there is an option to have these as local / relative links that would be great.

peterschrammel replied:
We should definitely use relative links.

TGWDB replied:
There is no consistent way of linking.

I was working on another (draft, so very provisional) pull request (see https://github.com/diffblue/cbmc/pull/5781) and it was recommended to use links to the online documentation at diffblue.com.

Exact quote is:

> README.md references http://cprover.diffblue.com/, so I'd go with that in top-level files (README.md, COMPILING.md, CODING_STANDARD.md, and possibly any files to be added at that level).

So should I document with:
A. Links to the local documentation path (e.g. doc/html/index.html)
B. Links to the absolute location (e.g. http://cprover.diffblue.com/index.html)
C. Something else? (E.g. relative html links, although these will not work when .md files are used like this one?)

peterschrammel replied:
I think the confusion is that the top-level .md files are currently not compiled into cprover.diffblue.com, but only those in `doc` and the `README.md` files in the source directories.

TGWDB replied:
My rationale was that this is a top level .md file (similar to README.md) and so we should link to the diffblue.com URLs. Since this is not going to be compiled into the website, then I understand the current linking is correct?

The above said, if we wish to move this document to another location (or have another reason to change), let me know and I can redo the links as relative.

martin-cs commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,228 @@
+
+## goto-analyzer
+
+This provides a way to access and invoke various forms of analysis on
```
It's a bit more detailed and restricted than that.  `goto-analyzer` is a wrapper around the abstract interpretation ( http://cprover.diffblue.com/background-concepts.html#abstract_interpretation_section ) implementation ( well, implementations, see http://cprover.diffblue.com/group__analyses.html ) in CPROVER.  It allows you to configure what abstractions are used and what is then done with them (verification, display, simplification, etc.).  At the moment the best documentation is here : http://cprover.diffblue.com/goto__analyzer__parse__options_8h.html but I have writing a section for the manual on my TODO list.

TGWDB replied:
Incorporated these comments into an updated version (some text taken almost directly).

Let me know if you want to coordinate on writing the section for the manual (I'd love to learn more about this, and help write the section if I can).

martin-cs commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,228 @@
+## goto-cc
+
+This is a compiler that also adds goto program information to the compiled
+programs. Note that `goto-cc` is the compiler and linker while `goto-gcc` and
```
`goto-cc` is intended as a drop-in replacement for `gcc` so that you can re-use the same build infrastructure.  The key difference is that it builds `goto-modelt`s instead of executables ( somewhere @tautschnig had a patch to build both but it is not the default ).  Really important is that depending on the same of the binary, you get emulation of different compilers, such as the Visual Studio one and the CodeWarrior one.

TGWDB replied:
Pushed a rewrite that I hope reflects this.

martin-cs commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,228 @@
+Provides a variety of difference checks between two goto programs (produced
+by `goto-cc`). This invokes some of the cbmc tools to convert the goto
+program and then determine which functions are added/removed/changed.
+
```
@peterschrammel ?

peterschrammel replied:
Essentially a `diff` tool for goto-programs.

TGWDB replied:
Updated the text for this (and also jdiff).

martin-cs commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,228 @@
+## janalyzer
+
+This provides a way to access and invoke various forms of analysis on
+Java programs.
```
Yes; `janalyzer` is a fork of `goto-analyzer` with the JVM front-end added.

TGWDB replied:
Updated the text for this.

martin-cs commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,228 @@
+
+This is the main analysis engine the performs the analysis
+of Java files using bounded model checking. This is a
+Java version of the `cbmc` tool (more documentation for
```
Pedantry perhaps but to me a "java version" implies it is written in Java.  The point is that `jbmc` is an old fork of `cbmc` with the JVM front-end added.

peterschrammel replied:
Yes, it's a BMC tool for Java bytecode (similar to cbmc being a BMC tool for C).
It has been separated from CBMC in order to support separate compilation of C and Java frontends and tools. Also, JBMC's command line follows conventions of `java` so that programs can be analysed in the same way you would run them using the `java` executable. There also lots of additional options to support nondeterministic initialisation of variable size arrays, data structures and strings, control how to assert on exceptions etc.

TGWDB replied:
Updated text for this (and added some clear phrases from Peter).

martin-cs commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,228 @@
+Provides a variety of difference checks between two goto programs (produced
+by `goto-cc`). This invokes some of the `cbmc` tools to convert the goto
+program and then determine which functions are added/removed/changed.
+
```
Likewise, I believe (@peterschrammel ?) that `jdiff` is `goto-diff` forked and the JVM front-end added.

peterschrammel replied:
Yes, just a clone of goto-diff for use on Java bytecode input.

TGWDB replied:
Updated text (and also on goto-diff)

martin-cs commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,228 @@
+## memory-analyzer
+
+This is a wrapper program that provides a front end to `gdb` that adds some
+useful features related to the other goto utilities. Note that to use
```
It's a bit more than that.  `memory-analyzer` allows you to run an compiled binary, using `gdb` and when you are at the right point, it will build a harness that allows you to analyze the program /from that state/.

TGWDB replied:
Added text on this, hope it is clear(er) now.

martin-cs commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,228 @@
+
+This is an (Satisfiability Modulo Theory) SMT solver that is shipped
+with CProver. The SMT solver here is built on top of a Satisfiability
+(SAT) solver. The `smt2_solver` is designer to work with files, and
```
It is basically an SMT interface to the CPROVER back-end prover.

TGWDB replied:
@martin-cs Can you please clarify whether you are proposing a change to the text (and if so perhaps propose some new text), or whether you were making a comment to help us understand?

martin-cs replied:
I think more of the later but let's try the former too...

`smt2_solver` is an SMT (Satisfiability Modulo Theory) solver.  It parses SMT-LIB 2 format and uses CPROVER's internal bit-blasting solver (see `solvers/`) to resolve queries.

martin-cs reviewed (approved):
Basically yes but there are a few more bits and links that might be useful.

peterschrammel commented on TOOLS_OVERVIEW.md:
````diff
@@ -0,0 +1,228 @@
+janalyzer --help
+```
+
+## java-unit
````
Not really an executable for users, but developers.

peterschrammel commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,228 @@
+
+## symtab2gb
+
+This utility is to compile a cbmc symbols table (in JSON format) into a goto binary.
```
This is used to integrate external language frontends, e.g. https://github.com/diffblue/gnat2goto/

peterschrammel commented on TOOLS_OVERVIEW.md:
````diff
@@ -0,0 +1,228 @@
+```
+
+
+## unit
````
Things below are for developers.

TGWDB replied:
Added developer utilities section and listed the developer tools there.

tautschnig commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,254 @@
+- [Developer Utilities](#developer-utilities)
+    - [java-unit](#java-unit)
+    - [unit](#unit)
+    - [others (converter, driver, file_converter, java-converter)](#others)
```
This is great, and makes me wonder whether this should be linked (or included in?) the top-level README file so that people accessing the repository see it right away?

TGWDB replied:
Added link in README.md

tautschnig commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,254 @@
+This is a compiler designed to be able to be dropped in to an existing
+build process to replace the existing compiler. `goto-cc` is able to do
+normal compilation, but is designed to output goto programs (optionally
+in addition to the normal compiled program). Note that `goto-cc` is the
+compiler and linker while `goto-gcc` and `goto-ld` are symbolic links to
+`goto-cc`. The additional object code is used as the internal representation
```
I'm not sure the fact that these are symlinks matters the most to a user. Perhaps focus on the resulting behaviour? `goto-cc` behaves differently, depending on the file name: the command-line options (and ensuing behaviour) are modelled after the program that has the same name as the suffix, i.e., `gcc`, `ld`, etc. (we also support `cl`, `link`, `bcc`, `as`).

TGWDB replied:
This has been updated now to clarify the role of the different files and de-emphasise the symlinks.

tautschnig commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,254 @@
+- [goto Programs](http://cprover.diffblue.com/group__goto-programs.html)
+for information on goto programs and how they are used.
+
+Note that `goto-cc` can emulate GCC, Visual Studio, and CodeWarrior
+compilers.
```
Ah, see my notes about for symlinks.

tautschnig commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,254 @@
+
+Provides a variety of difference checks between two goto programs
+produced by `goto-cc` (essentially a `diff` tool for goto programs).
+This invokes some of the cbmc tools to convert the goto
+program and then determine which functions are added/removed/changed.
```
I don't think that any other tools are being invoked?

tautschnig commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,254 @@
+## goto-harness
+
+This is a tool for creating a harness around a (part of a) goto program that
+can then be analysed inside the harness. Harnesses can be either function
```
What does "inside the harness" mean?

tautschnig commented on TOOLS_OVERVIEW.md:
```diff
@@ -0,0 +1,254 @@
+Default behaviour is to only show failed test cases.
+
+
+### others
+
+These (`converter`, `driver`, `file_converter`, `java-converter`)
+are build time utilities for other tools and not designed
+to be run independently.
```
I'm not sure these should even be mentioned?

TGWDB replied:
Rationale was to have list of all binaries (built on Linux, so some on other platforms may be missing). Open to removing the section though.

tautschnig replied:
Understood, but I fear this might cause more confusion than doing good. Not a blocker from my point of view, leaving it to you to decide.

tautschnig reviewed (approved):
Thank you for adding all this documentation! One overall comment: would be be worth merging the sections janalyzer/goto-analyzer, jdiff/goto-diff, jbmc/cbmc, java-unit/unit?

TGWDB committed:
Add overview of the CProver tools

This adds the TOOLS_OVERVIEW.md file (and links to it
from README.md). The Tools Overview is an overview of all
the tools that are part of the CProver package. Each
overview includes main points on how to use the tool and
what the tool is meant to do. Also further documentation
links are provided to better understand the tool.
